source/game_sa/CWeapon.cpp

```diff
@@ -25,54 +14,104 @@ float& fPlayerAimScale = *(float*)0x8D6110;
+    ReversibleHooks::Install("CWeapon", "TargetWeaponRangeMultiplier", 0x73B380, &CWeapon::TargetWeaponRangeMultiplier);
+    ReversibleHooks::Install("CWeapon", "Initialise", 0x73B4A0, &CWeapon::Initialise);
+    ReversibleHooks::Install("CWeapon", "DoWeaponEffect", 0x73E690, &CWeapon::DoWeaponEffect);
+}
+
+// 0x73A300
 void CWeapon::InitialiseWeapons() {
-    plugin::Call<0x73A300>();
+    CWeaponInfo::Initialise();
+    CShotInfo::Initialise();
+    CExplosion::Initialise();
+    CProjectileInfo::Initialise();
+    CBulletInfo::Initialise();
+
+    bPhotographHasBeenTaken = false;
+    ms_bTakePhoto = false;
 }
 
-// Converted from cdecl void CWeapon::ShutdownWeapons(void) 0x73A330
+// 0x73A330
 void CWeapon::ShutdownWeapons() {
-    plugin::Call<0x73A330>();
+    CWeaponInfo::Shutdown();
+    CShotInfo::Shutdown();
+    CExplosion::Shutdown();
+    CProjectileInfo::Shutdown();
+    CBulletInfo::Shutdown();
+
+    ms_PelletTestCol.RemoveCollisionVolumes();
 }
 
-// Converted from cdecl void CWeapon::UpdateWeapons(void) 0x73A360
+// 0x73A360
 void CWeapon::UpdateWeapons() {
-    plugin::Call<0x73A360>();
+    CShotInfo::Update();
+    CExplosion::Update();
+    CProjectileInfo::Update();
+    CBulletInfo::Update();
 }
 
-// Converted from thiscall void CWeapon::Shutdown(void) 0x73A380
+// 0x73A380
 void CWeapon::Shutdown() 
 {
-#ifdef USE_DEFAULT_FUNCTIONS
-    plugin::CallMethod<0x73A380, CWeapon*>(this);
-#else
-    int weaponModelID1 = CWeaponInfo::GetWeaponInfo(m_nType, 1)->m_nModelId1;
+    int32_t weaponModelID1 = CWeaponInfo::GetWeaponInfo(m_nType, 1)->m_nModelId1;
+    int32_t weaponModelID2 = CWeaponInfo::GetWeaponInfo(m_nType, 1)->m_nModelId2;
+
     if (weaponModelID1 != -1)
         CModelInfo::ms_modelInfoPtrs[weaponModelID1]->RemoveRef();
-    int weaponModelID2 = CWeaponInfo::GetWeaponInfo(m_nType, 1)->m_nModelId2;
+    
     if (weaponModelID2 != -1)
         CModelInfo::ms_modelInfoPtrs[weaponModelID2]->RemoveRef();
+
     m_nType = WEAPON_UNARMED;
     m_nState = WEAPONSTATE_READY;
     m_nTotalAmmo = 0;
     m_nAmmoInClip = 0;
     m_nTimeForNextShot = 0;
-#endif
 }
 
-// Converted from thiscall void CWeapon::AddGunshell(CEntity *creator,CVector const&position,CVector2D const&direction,float size) 0x73A3E0
-void CWeapon::AddGunshell(CEntity* creator, CVector const& position, CVector2D const& direction, float size) {
-    plugin::CallMethod<0x73A3E0, CWeapon*, CEntity*, CVector const&, CVector2D const&, float>(this, creator, position, direction, size);
+// 0x73A3E0
+void CWeapon::AddGunshell(CEntity* creator, CVector& position, const CVector2D& direction, float size) {
+
+    if (!creator || !creator->GetIsOnScreen())
+        return;
+
+    CVector& posDiff = TheCamera.GetPosition() - creator->GetPosition();
+
+    if (posDiff.Magnitude() > 10.0f)
+        return;
+
+    CVector velocity;
+
+    velocity.Set(direction.x, direction.y, CGeneral::GetRandomNumberInRange(0.4f, 1.6f));
```

Comment:
Might as well use ctor
```suggestion
    CVector velocity(direction.x, direction.y, CGeneral::GetRandomNumberInRange(0.4f, 1.6f));
```